### ISSUE TITLE
crash on iPhoneXS (only) on fs related operations

### ISSUE THREAD

**Issue by DiegoNegri** [janeasystems/nodejs-mobile#132]
Hi,

We have crashes on iPhone XS ONLY (not on simulator neither on iPhone X) with iOS 12, latest react-native-nodejs when code has errors accessing file system.

To reproduce:
- download sample 'UseMultipleChannels' and add at beginning of main.js the following:

process.on('uncaughtException', (err) => {
    console.log('Caught exception');
});

try {
    // this is to fool browserify so it doesn't try (in vain) to install iconv.
    iconv_package = 'iconv';
    Iconv = require(iconv_package).Iconv;
} catch (E) {
    // node-iconv not present
    console.log('ERROR SHOULD BE CATCHED')
}

The app runs fine on Simulator and iPhoneX (logging 'ERROR SHOULD BE CATCHED', it crashes on iPhoneXS with SIGABRT: 
libc++abi.dylib: terminating with uncaught exception of type Js::JavascriptException 

Thanks for your support

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,

Thank you for reporting this.
Unfortunately, I'm currently unable to try to reproduce the error on the specific "iPhone XS" device.
Are the iOS versions exactly the same on both devices? (iOS 12.0 , non beta)

I'd suggest avoiding trying to require iconv or patching the code that does that in your project dependencies.

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo 

Thanks for your quick reply.

Yes, the iOS versions are the same, no beta installed.

Got your suggestion, i avoided to require iconv but unfortunately there are crashes later after app startuo, i'm trying to intercept require/fs accesses to see if i can workaround it.

Keep you posted

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo 

Building nodejs-mobile and related libs in debug i can see:

_deps/chakrasim/core/lib/Runtime/Language/JavascriptEsceptionOperators.cpp_ reaches line 1131 and throws exception:  _throw JavascriptException(threadContext, exceptionObject);_ 

But, only on iPhoneXS, the process crashes for 'uncaught exception'.

On simulator/iPhoneX the exception is catched in:
__deps/chakrasim/core/lib/Runtime/Language/InterpreterStackFrame.cpp_ line 6584 in:
        catch (const Js::JavascriptException& err)
        {
            // We are using C++ exception handling which does not unwind the stack in the catch block.
            // For stack overflow and OOM exceptions, we cannot run user code here because the stack is not unwind.
            exception = err.GetAndClear();
        }

 Maybe it's related to compiler options related to iPhoneXS arm architecture ?

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,

Thank you for details.
It might be related to differences in the CPU, since iPhone XS ships with the "Apple A12 Bionic" and iPhone X ships with the "Apple A11 Bionic", though it's hard to tell why this would affect catching the exceptions.

Running the unit tests might shed some light into which features are failing: https://github.com/janeasystems/nodejs-mobile/blob/7d74e5eafd08d939dc99c027bf9a4e74e36b8ca8/doc_mobile/TESTING.md#ios

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo 

launching test on iPhoneX run fine, but executing the same shell './tools/mobile-test/ios/prepare-ios-tests.sh' on iPhoneXS fails as below:
...
(lldb)     script fruitstrap_device_app="/private/var/containers/Bundle/Application/4646B152-7794-4000-B32B-108DDA13D749/testnode.app"
(lldb)     script fruitstrap_connect_url="connect://127.0.0.1:56215"
(lldb)     target modules search-paths add /usr "/Users/diego.negri/Library/Developer/Xcode/iOS DeviceSupport/12.0 (16A5366a)/Symbols/usr" /System "/Users/diego.negri/Library/Developer/Xcode/iOS DeviceSupport/12.0 (16A5366a)/Symbols/System" "/private/var/containers/Bundle/Application/4646B152-7794-4000-B32B-108DDA13D749" "/Users/diego.negri/Documents/Products/nodejs-mobile-test/nodejs-mobile/out/ios.release/Release-iphoneos" "/var/containers/Bundle/Application/4646B152-7794-4000-B32B-108DDA13D749" "/Users/diego.negri/Documents/Products/nodejs-mobile-test/nodejs-mobile/out/ios.release/Release-iphoneos" /Developer "/Users/diego.negri/Library/Developer/Xcode/iOS DeviceSupport/12.0 (16A5366a)/Symbols/Developer"
**_(lldb)     command script import "/tmp/C2165CD7-B5D5-47BA-94F3-ED1620106EB8/fruitstrap_00008020-000E31323428002E.py"
error: module importing failed: ('invalid syntax', ('temp.py', 1, 27, 'import fruitstrap_00008020-000E31323428002E\n'))
backtrace unavailable_**

Have you any hint ?

**Comment by jaimecbernardo** [comment]
Thank you for giving it a try.
It's an `ios-deploy` issue with the iPhone XS device IDs starting having a dash now, it seems: https://github.com/ios-control/ios-deploy/issues/361
A PR seems to have been opened to solve that issue in `ios-deploy`.

**Comment by yunomeow** [comment]
Hi, I also have similar issue when using iPhoneXS and iPhoneXS Max.
It will crash with the following code.

```javascript
function throwException() {
    throw new Error('Exception');
}
try{
    throwException();
}catch (e){
    console.log(e);
}
```
It will not crash on older devices like iPhone7 plus with iOS 12.

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo, 

Debugging chackracore on iPhoneXS (armv8a) i can see catch management 'stop' propagating when a pure asm function is invoked ('arm64_CallFunction'), the problem is not visible on iPhone (armv8).

Documentation of armv8a talks about a new exception handling mechanism.

Have you any hint on solution/workaround ?

Thx
Diego

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo 

I got it working on iPhoneXS (and iPhoneX) changing chackracore invocation of js function.

The change just replaces only an asm function used to invoke variadic method with direct invocation of the method in c++.

If it's fine for you i'm going to open a PR.

Thx
Diego

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,

Thank you for the updates.
The new Apple CPU on iPhone XS is a ARMv8.3-A CPU, so it might have been affected by exception handling changes in the architecture. This means a fix to the assembly function is probably what's needed.
Could you please link to the changes you found in the documentation?

Good to hear your workaround is working.

You seem to have opened the PR to the wrong repository, by the way. (You opened the PR to upstream https://github.com/nodejs/node-chakracore )

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo ,

Follow the change i did in the forked project (sorry for PR on wrong repo):

in file: deps/chakrashim/core/lib/Runtime/Library/JavascriptFunction.cpp

removed line 1422:  

varResult = arm64_CallFunction((JavascriptFunction*)function, args.Info, args.Values, entryPoint);

and added a 'direct' entrypoint invocation like:

unsigned count = args.Info.Count;
        switch(args.Info.Count) 
        {
            case 0:
                varResult = CALL_ENTRYPOINT(function->GetScriptContext()->GetThreadContext(),
                    entryPoint, (JavascriptFunction*)function, args.Info);
                break;
            case 1:
                varResult = CALL_ENTRYPOINT(function->GetScriptContext()->GetThreadContext(),
                    entryPoint, (JavascriptFunction*)function, args.Info, args.Values[0]);
                break;
            case 2:
                varResult = CALL_ENTRYPOINT(function->GetScriptContext()->GetThreadContext(),
                    entryPoint, (JavascriptFunction*)function, args.Info, args.Values[0], args.Values[1]);
                break;
            case 3:
                varResult = CALL_ENTRYPOINT(function->GetScriptContext()->GetThreadContext(),
                    entryPoint, (JavascriptFunction*)function, args.Info, args.Values[0], args.Values[1], args.Values[2]);
                break;
...

This avoid the ASM implementation in 'arm64_CallFunction' and let exceptions flowing up.

**Comment by jaimecbernardo** [comment]
Thank you, here's a direct link for the commit with the proposed fix in the PR's repo: https://github.com/EidosMedia/nodejs-mobile/commit/9c2079567d33374a2860e214c6dbd90897817453

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,
Thank you for the contribution, but the proposed change causes a regression.
This test no longer passes, causing an access violation: https://github.com/janeasystems/nodejs-mobile/blob/7d74e5eafd08d939dc99c027bf9a4e74e36b8ca8/test/parallel/test-punycode.js

It can't be merged without further investigation.

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo ,

Debugging the failed test (test-punycode.js) i verified the failure is due to limited number of parameter supported by the patch (it was 50). Increasing this limit the test did succeed.
I'm going to run the full test suite before resubmitting the PR.

(by the way, of course, feel free to substitute the patch with an alternative more elegant implementation)

Thx 
Diego

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo ,

you can find a commit increasing the number of parameters supported (up to 256) at https://github.com/EidosMedia/nodejs-mobile/commit/d4f36d14bfb4740c04d62bc2fd7d73530a0d021e

Test run fine (parallel sequential message)

Please let me know for any feedback/info

Thx
Diego

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo ,

one more info, testing in our app i verified 'out of stack' error which i solved increasing stacksize of node thread in 'nodejs-mobile-react-native' (RNNodeJsMobile.m line 136).

Thanks for your support,
Diego

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,
Thank you for the contribution!
The extra stack space needed is an unfortunate side-effect, but it's something that's needed while the fix isn't done on the assembly side.
I'll give the solution a try.

**Comment by jaimecbernardo** [comment]
Hi @DiegoNegri ,
The tests have run successfully after increasing the test application's stack space.

Thank you for the contribution. I've opened the a PR with the code for integrating the code in this project: https://github.com/janeasystems/nodejs-mobile/pull/141

I think you can close the one you opened on https://github.com/nodejs/node-chakracore/pull/604

**Comment by DiegoNegri** [comment]
Hi @jaimecbernardo ,

glad it's fine, i closed [nodejs@604](https://github.com/nodejs/node-chakracore/pull/604) as you suggested.

**Comment by jaimecbernardo** [comment]
Thank you, @DiegoNegri .
The PR https://github.com/janeasystems/nodejs-mobile/pull/141 was merged in https://github.com/janeasystems/nodejs-mobile/commit/bb4b8548fc5f9b2ef5c4c2b450a61c68ea27a500

**Comment by jaimecbernardo** [comment]
Hi,

A fix to the `arm64_CallFunction` assembly was pushed: https://github.com/janeasystems/nodejs-mobile/commit/9e0b148670c990acf17bc503b430a6d78a2eeaa9

This will make the stack usage more efficient again.

**Comment by DiegoNegri** [comment]
Hi,
got it, great, thanks for the change

**Comment by nokai** [comment]
Hello

Is this problem fixed? The same thing happens to me now in the Iphone XS Max.

libc ++ abi.dylib: terminating with uncaught exception of type Js :: JavascriptException

If it's fixed, .. Where can I download the latest NodeMobile update?

**Comment by DiegoNegri** [comment]
Hi, 

the fix (latest refactor) is available since nodejs-mobile-v0.2.1

**Comment by jaimecbernardo** [comment]
Hi @nokai ,
The releases for the native library are here : https://github.com/JaneaSystems/nodejs-mobile/releases
